_posts/2020-03-16-dask-usage-at-blue-yonder.markdown

```diff
@@ -1,131 +1,131 @@
+  A source dataset is read, computations are performed on the data, and the result is written to
+  a target dataset.
+  In many cases, the layout of the source dataset (the partitioning, i.e., what data resides in which blob)
+  is used for parallelization. This means the algorithms work independently on the individual
+  blobs of the source dataset. Therefore our respective Dask graphs again look very simple, with parallel nodes each performing
+  the sequential operations of reading in the data from a source dataset blob, doing some computation on it,
+  and writing it out to a target dataset blob. Again, there is a final reduction node writing the target
+  dataset metadata.
+
+  ### Dataset maintenance
+
+  We also use Dask/Distributed for performing maintenance operations like garbage collection on datasets.
+  Garbage collection is necessary since we use a lazy approach when updating datasets and only delete or update
+  references to blobs in the dataset metadata, deferring the deletion of no longer used blobs.
+  In this case, Dask is used to parallelize the delete requests to the storage provider (Azure blob service).
+  In a similar fashion, we use Dask to parallelize copy operations for backing up datasets.
+
+  ### Repartitioning a dataset
+
+  Our algorithms rely on source datasets that are partitioned in a particular way. It is not always
+  the case that we have the data available in a suitable partitioning, for instance, if the data results
+  from a computation that used a different partitioning. In this case, we have to repartition the data.
+  This works by reading the dataset as a Dask distributed dataframe (ddf), repartitioning the ddf using
+  network shuffle, and writing it out again to a dataset.
+
+  ## Dask cluster setup at Blue Yonder
+
+  We run a somewhat unique setup of Dask clusters that is partly driven by the specific requirements
+  of our domain and partly historically grown. We are running a large number of individual clusters
+  that are homogeneous in size with many of the clusters dynamically scaled. The peculiarities of this
+  setup have in some cases triggered edge cases and uncovered bugs, which lead us to submit a number
+  of upstream pull requests.
+
+  ### Cluster separation
+
+  We are running separate Dask clusters for separate customers in order to isolate the customers from each other.
+  For the same reason, we also run separate clusters for production, staging, and development environments. 
+  But it does not stop there. For performance reasons, we install the Python packages holding the code needed
+  for the computations on each worker. Since the different components of our compute pipeline are implemented
+  and maintained as completely separate software packages whose requirements are not synchronized, this also
+  means we have to run a separate Dask cluster for each of the components. This leads to us operating more than
```

Comment:
I don't think it is easy to understand why we need to have different clusters are component level also.

Reply:
typo *at

Reply:
Yes, it's a bit tricky. Maybe an example could help here?